multicluster/controllers/multicluster/member/clusterset_controller.go

```diff
@@ -98,6 +98,10 @@ func (r *MemberClusterSetReconciler) Reconcile(ctx context.Context, req ctrl.Req
 		}
 		klog.InfoS("Received ClusterSet delete", "clusterset", req.NamespacedName)
 		if r.remoteCommonArea != nil {
+			klog.InfoS("Clean up all exported resources created by Antrea Multi-cluster controller", "clusterset", req.NamespacedName)
+			if err := r.deleteResourceExports(ctx); err != nil {
```

Comment:
Should we move these to stale controller too to be consistent with other cleanup? Or keeping them here you think we need not to worry about synchronization between stale controller and other reconcilers?

How about cleanup in the leader after the memberAnnounce is deleted? It is now done by stale controller. I would like the same pattern for both. I can see cleanup in the Announce or ClusterSet controller may remove the synchronization issue here, but how about the restart case - can there be a race condition between stale controller and other reconcilers? What is our solution to address that? Could we answer all these questions before changing the implementation again. I feel we just keep changing without good understanding of the problems.

Reply:
After refinement, ClusterSet controller will only try to clean up local auto-generated resources in the member, and it will delete the MemberClusterAnnounce as well. The ResourceExports cleanup will be done in the stale controller in the leader cluster.
Previous stale controller's race condition bug was fixed in this PR https://github.com/antrea-io/antrea/pull/4853.
The new stale controller in the member is still calling the same clean up process when there is a ClusterSet. And if there is no ClusterSet CR, it will only clean up local resources during restart, which should have no race condition.

When the member is crashed during deleting MemberClusterAnnounce, there will be stale one in the leader cluster. It will be cleaned up after 2 hours in the leader cluster, then corresponding ResourceExports will be removed too.
If the member is back and a new ClusterSet is created in 2 hours, the member will update the timestamp of  MemberClusterAnnounce again, and the stale controller in member will check stale ResoureExports and stale local resources and clean them.

Reply:
For the race condition between stale controller and other controllers, I meant the following cases:

1. Member stale controller can call `cleanUpResourcesCreatedByMC` at startup. How if a new ClusterSet is created when that is going. Could `cleanUpResourcesCreatedByMC` delete new resources created for the new ClusterSet?
2. Leader stale controller can call `cleanUpStaleResourceExports` at startup. How if a member announce is recreated? Could `cleanUpStaleResourceExports` delete new Exports created by the re-joined member?
3. Leader stale controller `Reconcile` can delete `ResourceExports` of a deleted member announce. How if the member announce is being updated (maybe the update will fail with DeletionTimestamp?)?

Reply:
I am thinking to add a finalizer at member cluster's ClusterSet to avoid the cases you mentioned.
If there is a finalizer at member's ClusterSet, it helps to handle the deletion failure issue when the controller crashed and no need to add periodical job to delete memberclusterannounce since it will be guaranteed to delete by member controller. 
When a ClusterSet is with finalizer and being deleted but clean up is not fully done and finalizer is not removed, the new ClusterSet will fail to create by webhook because of the existing ClusterSet. This can help to avoid the second/third cases. A new MemberClusterAnnounce won't be created as well before clean up is done.

Let me know what's your suggestion. Thanks.

Reply:
> and no need to add periodical job to delete memberclusterannounce since it will be guaranteed to delete by member controller.
I do not understand this. I think you want to delete memberAnnounce when a member is disconnected for long and so added the periodical job.

Finalizer should not be the only way to solve the race conditions (e.g. for #1 and #2 I feel we can use a lock too). I personally do not like to depend on an external mechanism when it is not necessary. But what is the current usage of Finalizer in MC Controller? Do we use that for similar cases already?